Add Upnp volume control/status to SamsungTV

epenet opened the pull request:
https://github.com/home-assistant/core/pull/68717 needs to be merged first

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add Upnp volume control/status to SamsungTV
Adds `SUPPORT_VOLUME_SET` to `_attr_supported_features`

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

epenet committed:
Add Upnp to SamsungTV

epenet commented:
cc @bdraco, @sermayoral

I still need to add tests, and parse the source list.
However, since GetSourceList isn't supported on my TV I cannot complete it.

epenet committed:
Default UpnpFactory to fail

epenet committed:
Add logs for set_volume_level

epenet committed:
Add tests

epenet committed:
Add TODO

epenet committed:
Remove upnp app_list request

epenet commented:
> cc @bdraco, @sermayoral
> 
> I still need to add tests, and parse the source list. However, since GetSourceList isn't supported on my TV I cannot complete the source list just yet.

I have removed the Upnp app_list request - I can add in a follow-up PR
I think this is good to go

sermayoral commented:
@epenet. What do you need?

epenet commented:
> @epenet. What do you need?

For this PR => just confirmation that these three work correctly for you:
- get volume level
- set volume level
- get mute status

epenet commented:
> @epenet. What do you need?

For Upnp app-list, maybe you can ping me on discord (https://discord.com/invite/home-assistant / epenet#1188) to discuss further

sermayoral commented:
> For Upnp app-list, maybe you can ping me on discord (https://discord.com/invite/home-assistant - epenet#1188) to discuss further

Not A Number:

![image](https://user-images.githubusercontent.com/30116095/160128745-32a91089-7cce-4b0d-86a8-a3c422807427.png)

Here is the full log: https://pastebin.com/rV1HyFx9

epenet commented:
@sermayoral I have set a default value for the TVs that do not respond on url `http://{self._host}:9197/dmr`

I am not sure how to handle TVs that have Upnp on a different URL.

bdraco commented on homeassistant/components/samsungtv/media_player.py:
```diff
@@ -179,21 +193,72 @@ async def async_update(self) -> None:
+            upnp_factory = UpnpFactory(upnp_requester)
+            with contextlib.suppress(UpnpConnectionError):
+                self._upnp_device = await upnp_factory.async_create_device(
+                    f"http://{self._host}:9197/dmr"
```
We should be able to get the port from `async_step_ssdp`

bdraco replied:
I think we also need to shutdown the `self._upnp_device` in `async_will_remove_from_hass`

bdraco commented:
We need to add `"async-upnp-client==0.27.0"` to `manifest.json`

bdraco committed:
adjust

bdraco committed:
adjust

bdraco commented:
```
  "ssdp": [
    {
      "st": "urn:samsung.com:device:RemoteControlReceiver:1"
    },
    {
      "manufacturer": "Samsung",
      "st": "urn:schemas-upnp-org:service:RenderingControl:1"
    },
    {
      "manufacturer": "Samsung Electronics",
      "st": "urn:schemas-upnp-org:service:RenderingControl:1"
    }    
  ],
```

ssdp needs to change as well

bdraco committed:
adjust

bdraco commented:
@sermayoral does the last push fix the uPnP url for your device ?

bdraco committed:
cover

bdraco committed:
fixes

sermayoral commented:
> @sermayoral does the last push fix the uPnP url for your device ?

@bdraco gonna try right now

sermayoral commented:
@bdraco try this as a custom component got the following error:

```
Logger: aiohttp.server
Source: requirements.py:180
First occurred: 23:27:39 (1 occurrences)
Last logged: 23:27:39

Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_protocol.py", line 435, in _handle_request
    resp = await request_handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_app.py", line 504, in _handle
    resp = await handler(request)
  File "/usr/local/lib/python3.9/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 60, in security_filter_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 100, in forwarded_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 28, in request_context_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 79, in ban_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 219, in auth_middleware
    return await handler(request)
  File "/usr/src/homeassistant/homeassistant/components/http/view.py", line 137, in handle
    result = await result
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 134, in post
    return await super().post(request)
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 62, in wrapper
    result = await method(view, request, *args, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 72, in post
    result = await self._flow_mgr.async_init(
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 203, in async_init
    flow, result = await task
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 221, in _async_init
    flow = await self.async_create_flow(handler, context=context, data=data)
  File "/usr/src/homeassistant/homeassistant/config_entries.py", line 740, in async_create_flow
    await async_process_deps_reqs(self.hass, self._hass_config, integration)
  File "/usr/src/homeassistant/homeassistant/setup.py", line 360, in async_process_deps_reqs
    await requirements.async_get_integration_with_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 83, in async_get_integration_with_requirements
    await _async_process_integration(hass, integration, done)
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 99, in _async_process_integration
    await async_process_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 162, in async_process_requirements
    await _async_process_requirements(
  File "/usr/src/homeassistant/homeassistant/requirements.py", line 180, in _async_process_requirements
    raise RequirementsNotFound(name, [req])
homeassistant.requirements.RequirementsNotFound: Requirements for samsungtv not found: ['async-upnp-client==0.27.0'].
```

bdraco commented:
@sermayoral I'm assuming you are not running dev which has a newer async-upnp-client so the version you have is conflicting but the older version should work just fine

If you drop the requirements line `async-upnp-client==0.27.0` in `manifest.json` it should work

sermayoral commented:
@bdraco I'm running stable branch. I've upgraded the async-upnp-client library in the docker container, but it still doesn't work. Same as https://github.com/home-assistant/core/pull/68663#issuecomment-1079024958

Here is the log: https://pastebin.com/FMsaGSZ1

bdraco commented:
> 

What does the config entry data look like in `.storage/core.config_entries`?

bdraco commented:
```
            {
                "entry_id": "0d3be2d2247e1c5515730c7f7c7f67ea",
                "version": 2,
                "domain": "samsungtv",
                "title": "89 master bed tv (QN55Q6DAAFXZA)",
                "data": {
                    "host": "192.168.106.51",
                    "manufacturer": "Samsung",
                    "method": "websocket",
                    "model": "QN55Q6DAAFXZA",
                    "name": "89 master bed tv",
                    "port": 8002,
                    "ssdp_location": "http://192.168.106.51:9197/dmr"
                },
                "options": {},
                "pref_disable_new_entities": false,
                "pref_disable_polling": false,
                "source": "zeroconf",
                "unique_id": "82becf53-fae5-4e0b-a125-64b5f08c04db",
                "disabled_by": null
            },
```
You should have an `ssdp_location` now

bdraco commented:
> @bdraco I'm running stable branch. I've upgraded the async-upnp-client library in the docker container, but it still doesn't work. Same as [#68663 (comment)](https://github.com/home-assistant/core/pull/68663#issuecomment-1079024958)

You likely have to remove the requirement from the custom samsungtv `manifest.json` since the other integrations will downgrade it again

sermayoral commented:
> > 
> 
> What does the config entry data look like in `.storage/core.config_entries`?

```
            {
                "entry_id": "2bd9715936db8016de2aeb9cb0e97474",
                "version": 2,
                "domain": "samsungtv",
                "title": "[TV]Samsung LED50 SYM (UE50H6400)",
                "data": {
                    "host": "192.168.33.32",
                    "mac": "24:4b:03:41:42:7d",
                    "manufacturer": "Samsung Electronics",
                    "method": "encrypted",
                    "model": "UE50H6400",
                    "name": "[TV]Samsung LED50 SYM",
                    "port": 8000,
                    "token": "05561012595d2b187ca96d5747f7ab3d",
                    "session_id": "1"
                },
                "options": {},
                "pref_disable_new_entities": false,
                "pref_disable_polling": false,
                "source": "ssdp",
                "unique_id": "068e7780-006e-1000-bc6f-244b0341427d",
                "disabled_by": null
            }
            ```

bdraco commented:
O wait you have two different ssdp_location in your discovery

bdraco committed:
Only set ssdp_location with the correct st

bdraco committed:
Merge branch 'dev' into samsungtv-upnp

bdraco commented:
I added a guard to make sure the ssdp_location only gets set when the ssdp discovery has the correct ST

bdraco commented:
@sermayoral Should be good to test again (will still need the requirement removed)

bdraco commented:
If its working correctly `ssdp_location` should appear in the config entry data

sermayoral commented:
@bdraco still not working :-(

manifest.json:
```
  "requirements": [
    "getmac==0.8.2",
    "samsungctl[websocket]==0.7.1",
    "samsungtvws[async,encrypted]==2.5.0",
    "wakeonlan==2.0.1"
  ],
```

config entry data:
```
            {
                "entry_id": "7d6db4fba18936d11387480ed297e55f",
                "version": 2,
                "domain": "samsungtv",
                "title": "[TV]Samsung LED50 SYM (UE50H6400)",
                "data": {
                    "host": "192.168.33.32",
                    "mac": "24:4b:03:41:42:7d",
                    "manufacturer": "Samsung Electronics",
                    "method": "encrypted",
                    "model": "UE50H6400",
                    "name": "[TV]Samsung LED50 SYM",
                    "port": 8000,
                    "token": "1d67d9c0c63c176a9188dd54ffcf4500",
                    "session_id": "2"
                },
                "options": {},
                "pref_disable_new_entities": false,
                "pref_disable_polling": false,
                "source": "ssdp",
                "unique_id": "068e7780-006e-1000-bc6f-244b0341427d",
                "disabled_by": null
            }
```

Logs: https://pastebin.com/9HmwiS7V

bdraco commented:
Its definitely getting the discovery.  I'm not sure why its not updating the config entry though

sermayoral commented:
> Its definitely getting the discovery. I'm not sure why its not updating the config entry though

The log shows it: `ssdp_location='http://192.168.33.32:7676/smp_15_'`, and this is right. This url contains de following: https://pastebin.com/tfiY3zJk

bdraco committed:
additional test

bdraco committed:
add some more debug

bdraco commented:
Added some more debug. Maybe it will show whats going on

sermayoral commented:
@bdraco here is the log: https://pastebin.com/2x3FYS1j

bdraco commented:
`2022-03-26 00:39:10 DEBUG (MainThread) [custom_components.samsungtv] Set SSDP location to: http://192.168.33.32:7676/smp_15_`

So its finding it but its never getting into the config entry

sermayoral commented:
> `2022-03-26 00:39:10 DEBUG (MainThread) [custom_components.samsungtv] Set SSDP location to: http://192.168.33.32:7676/smp_15_`
> 
> So its finding it but its never getting into the config entry

I don't understand...
```
            {
                "entry_id": "ef34a8a804b99224987266b26b2baa2c",
                "version": 2,
                "domain": "samsungtv",
                "title": "[TV]Samsung LED50 SYM (UE50H6400)",
                "data": {
                    "host": "192.168.33.32",
                    "mac": "24:4b:03:41:42:7d",
                    "manufacturer": "Samsung",
                    "method": "encrypted",
                    "model": "UE50H6400",
                    "name": "[TV]Samsung LED50 SYM",
                    "port": 8000,
                    "token": "887960e8a3fc1dfe77eadcc9e6947299",
                    "session_id": "3"
                },
                "options": {},
                "pref_disable_new_entities": false,
                "pref_disable_polling": false,
                "source": "user",
                "unique_id": "068e7780-006e-1000-bc6f-244b0341427d",
                "disabled_by": null
            }
```

bdraco committed:
more debug

bdraco commented:
I'm going to try deleting mine and setting it up from scratch to see if I can get it to not add the ssdp location

bdraco commented:
Unfortunately that worked just fine

bdraco committed:
fix race

bdraco commented:
@sermayoral  I think its a race between the two SSDP flows and the one with the ssdp_location we want loses the race for your setup.

I just pushed a fix for that *I hope*

sermayoral commented:
> @sermayoral I think its a race between the two SSDP flows and the one with the ssdp_location we want loses the race for your setup.
> 
> I just pushed a fix for that _I hope_

Mmmm wait. I have my custom component working in other Home Assistant Instance (other PC). Maybe its SSDP is hurting our tests...

bdraco commented:
> Mmmm wait. I have my custom component working in other Home Assistant Instance (other PC). Maybe its SSDP is hurting our tests...

I don't think it will affect it

sermayoral commented:
Same @bdraco:

```
            {
                "entry_id": "69057d5d55f5fc3a817137a6da175b90",
                "version": 2,
                "domain": "samsungtv",
                "title": "[TV]Samsung LED50 SYM (UE50H6400)",
                "data": {
                    "host": "192.168.33.32",
                    "mac": "24:4b:03:41:42:7d",
                    "manufacturer": "Samsung",
                    "method": "encrypted",
                    "model": "UE50H6400",
                    "name": "[TV]Samsung LED50 SYM",
                    "port": 8000,
                    "token": "375e637e83372b4f10a9d1794aa1a4e6",
                    "session_id": "4"
                },
                "options": {},
                "pref_disable_new_entities": false,
                "pref_disable_polling": false,
                "source": "user",
                "unique_id": "068e7780-006e-1000-bc6f-244b0341427d",
                "disabled_by": null
            }
```

Log: https://pastebin.com/1U2f4PmT

bdraco committed:
fix wrong port for encrypted websocket

bdraco committed:
reorder again

bdraco committed:
one more place it could fail

bdraco committed:
one more place it could fail

bdraco commented:
@sermayoral by change is there a reauth flow pending ?

sermayoral commented:
> by change is there a reauth flow pending ?

In each test I delete the component, and after adding the new version, I have to re-authenticate again. I don't know if that's what you mean

bdraco committed:
remove encrypted check since it does not actually do anything and is the same as the websocket check

bdraco commented on homeassistant/components/samsungtv/bridge.py:
```diff
@@ -90,14 +90,6 @@ async def async_get_device_info(
         if info := await bridge.async_device_info():
             return port, METHOD_WEBSOCKET, info
 
-    # Try encrypted websocket port
-    bridge = SamsungTVBridge.get_bridge(
-        hass, METHOD_ENCRYPTED_WEBSOCKET, host, ENCRYPTED_WEBSOCKET_PORT
-    )
-    result = await bridge.async_try_connect()
-    if result == RESULT_SUCCESS:
-        return port, METHOD_ENCRYPTED_WEBSOCKET, await bridge.async_device_info()
-
```
I think this check need to come before the non-encrypted check and see if this url works https://github.com/xchwarze/samsung-tv-ws-api/blob/master/samsungtvws/encrypted/remote.py#L102

bdraco committed:
fix encrypted detection

bdraco commented on homeassistant/components/samsungtv/bridge.py:
```diff
@@ -84,19 +86,23 @@ async def async_get_device_info(
+        hass, METHOD_WEBSOCKET, host, WEBSOCKET_NO_SSL_PORT
+    )
+    if info := await bridge.async_device_info():
+        encrypted_bridge = SamsungTVEncryptedBridge(
+            hass, METHOD_ENCRYPTED_WEBSOCKET, host, ENCRYPTED_WEBSOCKET_PORT
+        )
+        if await encrypted_bridge.async_can_connect():
```
async_try_connect on the encrypted broke the legacy tv discovery so I added `async_can_connect` but it needs to be moved into the upstream lib

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
fix another race

bdraco committed:
fix encrypted setup

bdraco committed:
dry

bdraco committed:
fix timeout

bdraco committed:
comment

bdraco committed:
reduce number of requests to the tv to fix overload

bdraco committed:
more tweaks

bdraco committed:
more tweaks

bdraco committed:
fixes

bdraco committed:
fixes

bdraco committed:
fix

bdraco committed:
fix legacy discovery

bdraco commented:
legacy, websocket, encrypted websocket all working for me now with this branch

@sermayoral can you give it another shot ?

bdraco commented:
Protest still need to be fixed, but I think we can split the config flow and bridge changes into a separate PR

bdraco commented:
And there is also the matter of merging the encrypted check to the upstream lib

bdraco commented:
I think we can actually avoid the upstream changed by keeping it guarded in the websocket to check

bdraco committed:
avoid the need for an upstream change

bdraco committed:
reduce

bdraco committed:
dry

bdraco committed:
dry

bdraco committed:
test fixes

bdraco committed:
fix tests

bdraco committed:
more fixes

bdraco committed:
coverage

bdraco commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -286,6 +373,9 @@ def _abort_if_manufacturer_is_not_samsung(self) -> None:
         if not self._manufacturer or not self._manufacturer.lower().startswith(
             "samsung"
         ):
+            LOGGER.debug(
+                "Aborting flow since manufacturer:%s is not samsung", self._manufacturer
+            )
```
```suggestion
```

bdraco committed:
Update homeassistant/components/samsungtv/config_flow.py

bdraco committed:
reduce

bdraco committed:
Merge branch 'samsungtv-upnp' of github.com:epenet/home-assistant-core into samsungtv-upnp
t push epenet samsungtv-upnp

bdraco commented on homeassistant/components/samsungtv/config_flow.py:
```diff
@@ -115,48 +133,73 @@ async def _async_set_unique_id_from_udn(
+        assert method is not None
+        LOGGER.debug("Try connect determined method to use: %s", method)
+        self._bridge = SamsungTVBridge.get_bridge(self.hass, method, self._host)
+        if self._bridge.method != METHOD_WEBSOCKET:
```
I still don't like this since it only gives you one chance to hit the button on the websocket tv

bdraco committed:
ensure they can try again if they miss the popup

bdraco committed:
fix retry

bdraco committed:
cover

bdraco commented:
Testing looks good

I'll split this after some sleep

bdraco committed:
Merge branch 'dev' into samsungtv-upnp

bdraco committed:
Add support for setting up encrypted samsung tvs from config flow

A reauth flow is no longer required to detect the need for
encryption as it can now be detected during setup

Fixes being able to setup legacy TVs since the encrypted
check previously would prevent them from proceeding

Improves the failure case where the user cannot hit
the button to allow Home Assistant connect to the TV
in time by allowing them to retry

bdraco committed:
Update tests/components/samsungtv/__init__.py

bdraco committed:
Update tests/components/samsungtv/__init__.py

bdraco commented:
Changes split into https://github.com/home-assistant/core/pull/68717

bdraco committed:
isort

bdraco committed:
rename ssdp key since we will need one for main_tv_agent as well

bdraco committed:
Merge branch 'samsungtv_encrypted_flow' into samsungtv-upnp

bdraco committed:
rename for merge fix

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into samsungtv-upnp

bdraco committed:
one more conflict

bdraco committed:
guards

bdraco commented:
Fixed the conflicts

epenet commented on homeassistant/components/samsungtv/media_player.py:
```diff
@@ -104,6 +115,10 @@ def __init__(
         self._config_entry = config_entry
         self._host: str | None = config_entry.data[CONF_HOST]
         self._mac: str | None = config_entry.data.get(CONF_MAC)
+        self._ssdp_location = (
+            config_entry.data.get(CONF_SSDP_RENDERING_CONTROL_LOCATION)
+            or f"http://{self._host}:9197/dmr"
+        )
```
Should we drop the default value, and instead make SUPPORT_VOLUME_SET dependant on the presence of CONF_SSDP_RENDERING_CONTROL_LOCATION?

bdraco replied:
done

bdraco committed:
do not set SUPPORT_VOLUME_SET if there is no ssdp location

bdraco committed:
do not set SUPPORT_VOLUME_SET if there is no ssdp location

bdraco committed:
gather cleanup

bdraco commented:
In testing found a case where reauth was requested when it wasn't needed.  If you turn off the tv during an update

```2022-03-27 11:28:29 INFO (MainThread) [homeassistant.components.samsungtv] Failed to get remote for 192.168.106.51, re-authentication required: ConnectionClosedError(None, Close(code=1011, reason='keepalive ping timeout'), None)
```

I'll open another PR for that

bdraco committed:
Merge branch 'dev' into samsungtv-upnp

MartinHjelmare commented on homeassistant/components/samsungtv/media_player.py:
```diff
@@ -179,21 +196,77 @@ async def async_update(self) -> None:
+            return
+
+        get_volume, get_mute = await asyncio.gather(
+            service.action("GetVolume").async_call(InstanceID=0, Channel="Master"),
```
I don't think this is in line with our design requirements. Protocol details should stay in a 3rd party library. The library should handle the UPnP setup and command details.

bdraco replied:
I figured this was ok since we have the similar pattern in the upnp and dlna* integrations

bdraco replied:
Maybe we create a new profile for Samsung Tv 

async_upnp_client.profiles.samsungtv?

MartinHjelmare replied:
I would consider them exceptions as they are general integrations similar to mqtt.

The Samsung TV integration is not that.

bdraco replied:
@epenet async_upnp_client.profiles.samsungtv or a new library?  It doesn't seem to fit in any of the existing ones.

Not sure what is better

bdraco replied:
I haven't looked into the details, and I won't be back home for a while so but I wonder if we could actually get subscriptions working as well so we could get push updates.
